text/0000-hash-simplification.md

```diff
@@ -0,0 +1,405 @@
+
+One of the current aspirations for the `std::hash` module was to be appropriate
+for hashing algorithms such as MD5, SHA\*, etc. The current API has proven
+inadequate, however, for the primary reason of hashing being so generic. For
+example it should in theory be possible to calculate the SHA1 hash of a byte
+slice via:
+
+```rust
+let data: &[u8] = ...;
+let hash = std::hash::hash::<&[u8], Sha1>(data);
+```
+
+There are a number of pitfalls to this approach:
+
+* Due to slices being able to be hashed generically, each byte will be written
+  individually to the `Sha1` state, which is likely to not be very efficient.
+* Due to slices being able to be hashed generically, the length of the slice is
+  first written to the `Sha1` state, which is likely not desired.
+
+The key observation is that the hash values produced in a Rust program are
+**not** reproducible outside of Rust. For this reason, APIs for reproducible
+hashes to be verified elsewhere will explicitly not be considered in the design
+for `std::hash`. It is expected that an external crate may wish to provide a
+trait for these hashing algorithms and it would not be bounded by
+`std::hash::Hash`, but instead perhaps a "byte container" of some form.
+
+# Detailed design
+
+This RFC considers two possible designs as a replacement of today's `std::hash`
+API. One is a "minor refactoring" of the current API while the
+other is a much more radical change towards being conservative. This section
+will propose the more radical change and the other may be found in the
+[Alternatives](#alternatives) section.
+
+## API
+
+The new API of `std::hash` will be:
+
+```rust
+trait Hash {
+    fn hash(&self) -> usize;
+}
+
+fn combine(a: usize, b: usize) -> usize;
+```
+
+The `Writer`, `Hasher`, and `SipHasher` structures/traits will all be removed
+from `std::hash`. This definition is more or less the Rust equivalent of the
+Java/C++ hashing infrastructure. This API is a vast simplification of what
+exists today and allows implementations of `Hash` as well as consumers of `Hash`
+to quite ergonomically work with hash values as well as hashable objects.
+
+> **Note**: The choice of `usize` instead of `u64` reflects [C++'s
+> choice][cpp-hash] here as well, but it is quite easy to use one instead of
+> the other.
+
+## Hashing algorithm
+
+With this definition of `Hash`, each type must pre-ordain a particular hash
+algorithm that it implements. Using an alternate algorithm would require a
+separate newtype wrapper.
+
+Most implementations will still use `#[derive(Hash)]` which will leverage
+`hash::combine` to combine the hash values of aggregate fields. Manual
+implementations which only want to hash a select number of fields would look
+like:
+
+```rust
+impl Hash for MyType {
+    fn hash(&self) -> usize {
+        // ignore field2
+        (&self.field1, &self.field3).hash()
+    }
+}
+```
+
+A possible implementation of combine can be found [in the boost source
+code][boost-combine].
```

Comment:
That function appears to be part of boost functional. I couldn't find anything on that page that mentioned security in any way which makes me think that that function might be designed to be fast without being secure against DOS attacks. It doesn't look like it would be particularly resistant to any attacks. Did I miss something where the boost developers discussed the security implications of that function?

Side note: I have absolutely no idea how to leave a comment on multiple lines of a diff.